homeassistant/components/mqtt/__init__.py

```diff
@@ -1048,3 +1060,27 @@ def device_info(self):
             info['via_hub'] = (DOMAIN, self._device_config[CONF_VIA_HUB])
 
         return info
+
+
+@websocket_api.async_response
+@websocket_api.websocket_command({
+    vol.Required('type'): 'mqtt/subscribe',
+    vol.Required('topic'): valid_subscribe_topic,
+})
+async def websocket_subscribe(hass, connection, msg):
+    """Subscribe to a MQTT topic."""
+    if not connection.user.is_admin:
+        raise Unauthorized
+
+    async def forward_messages(topic: str, payload: str, qos: int):
+        """Forward events to websocket."""
+        connection.send_message(websocket_api.event_message(msg['id'], {
```

Comment:
It would be really useful to forward the retained flag also to view and remove retained messages from the frontend.
But maybe that's for a different PR.

Reply:
how do we get the retained flag? it doesn't seem to be part of the subscription callback?

Reply:
Right, the subscription callback will have to be updated. Maybe better to leave it for a separate PR.